QuickLook sometimes shows broken images for image sources when loading web archives

whsieh opened the pull request:
#### 408858f600d2b7aafb86848018a1b8858f8dfd17
<pre>
QuickLook sometimes shows broken images for image sources when loading web archives
<a href="https://bugs.webkit.org/show_bug.cgi?id=272005">https://bugs.webkit.org/show_bug.cgi?id=272005</a>
<a href="https://rdar.apple.com/123918480">rdar://123918480</a>

Reviewed by Sihui Liu.

When previewing web archives, QuickLook (more precisely, `QLPreviewController`) loads the archive
data in a `WKWebView` that is configured to have an empty set of `_allowedNetworkHosts` (thereby
preventing network access when loading the archive). While this is a perfectly reasonable behavior
for the purposes of loading web archives, it exposes an issue when rendering images with sources
specified via the `srcset` attribute — if the &quot;best fit&quot; chosen for the image in QuickLook&apos;s web
view isn&apos;t the same source as the one that was chosen when creating the web archive, we end up with
a broken image since the selected source isn&apos;t present in the web archive, and we also can&apos;t
download it over the network. Since everything from viewport/screen dimensions to the device scale
factor can influence this, we can end up with broken images when copying images with `srcset`s in
many cases.

To mitigate this, we add some logic to detect when image sources are unavailable when loading web
archives (*and* we&apos;re additionally not allowed to fetch them from over the internet), and avoid
surfacing these image sources as potential candidates when selecting the best fit. See below for
more details.

Test: LoadWebArchive.PreferCachedImageSourceOverBrokenImage

* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::bestFitSourceFromPictureElement):
(WebCore::HTMLImageElement::selectImageSource):

Pass the new filtering function (using `ImageLoader::shouldIgnoreCandidateWhenLoadingFromArchive`
below) into `bestFitSourceForImageAttributes`.

* Source/WebCore/html/parser/HTMLSrcsetParser.cpp:
(WebCore::bestFitSourceForImageAttributes):

Add a new optional filtering function that takes an `ImageCandidate` and returns whether or not we
can even attempt to load the candidate; for now, this is `false` only in the case where we&apos;re
loading archived resources, and we&apos;re not allowed to load the URL over the network.

* Source/WebCore/html/parser/HTMLSrcsetParser.h:
(WebCore::bestFitSourceForImageAttributes):
* Source/WebCore/loader/DocumentLoader.h:
(WebCore::DocumentLoader::hasArchiveResourceCollection const):

Add a new helper function, which we can use to bail early in
`ImageLoader::shouldIgnoreCandidateWhenLoadingFromArchive`, to avoid eagerly parsing all URLs in the
`srcset`.

* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::shouldIgnoreCandidateWhenLoadingFromArchive const):
* Source/WebCore/loader/ImageLoader.h:
* Tools/TestWebKitAPI/Tests/mac/LoadWebArchive.mm:
(TestWebKitAPI::TEST(LoadWebArchive, PreferCachedImageSourceOverBrokenImage)):

Add a new API test to exercise the change, by loading a web archive from a web view at 1x device
pixel scale that contains an image with a `srcset`, and loading the archive in another web view at
2x scale. Before the changes in this patch, this results in a broken image.

* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView contentsAsWebArchive]):

Canonical link: <a href="https://commits.webkit.org/276952@main">https://commits.webkit.org/276952@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/41aaf8e4c5e208821bb23341148da10e3f166252

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46196 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25327 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48781 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48868 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42237 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48503 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29661 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22775 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37744 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46774 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22394 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39836 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18953 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19783 "layout-tests (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40934 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4240 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42491 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41283 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50670 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21192 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17681 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44929 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22492 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43844 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10240 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22851 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22186 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -677,4 +678,27 @@ VisibleInViewportState ImageLoader::imageVisibleInViewport(const Document& docum
+
+    Ref document = element().document();
+    RefPtr loader = document->loader();
+    if (!loader || !loader->hasArchiveResourceCollection())
```
Just to confirm, this will ensure that the new logic only applies to web archives, and not loaded web pages right?

(which is necessary, since `<source>` selection in general is not based on an ability to load the content)

However, if this only applies to web archives, maybe the method name needs to be more specific?

whsieh replied:
Yep, this logic is specific to web archives. I'll refactor this to make it more clear — e.g., `shouldIgnoreCandidateWhenLoadingFromArchive`

shallawa commented on Source/WebCore/html/HTMLImageElement.cpp:
```diff
@@ -302,7 +302,10 @@ ImageCandidate HTMLImageElement::bestFitSourceFromPictureElement()
         auto sourceSize = sizesParser.length();
 
-        candidate = bestFitSourceForImageAttributes(document().deviceScaleFactor(), nullAtom(), srcset, sourceSize);
+        candidate = bestFitSourceForImageAttributes(document().deviceScaleFactor(), nullAtom(), srcset, sourceSize, [&](auto& candidate) {
+            return m_imageLoader->shouldIgnoreCandidateWhenLoadingFromArchive(candidate);
+        });
```
This lambda will never be called because in `bestFitSourceForImageAttributes()`
```
if (srcsetAttribute.isNull())
```
will be always true in this case.

whsieh replied:
~Ah, good catch. I'll revert this part.~

Actually, I'm not so sure. This codepath is reachable in the case where the `srcset` attribute is non-empty, right? (In other words, we _wouldn't_ bail due to `srcsetAttribute.isNull()`).

(I'll keep this here for now)

szewai replied:
`srcsetAttribute`  is the third parameter

whsieh replied:
Hm... it's indeed the third parameter — unless I'm missing something, it seems we only get to this point if the `srcset.isEmpty()` check fails above, so we should _never_ bail here, when called from `bestFitSourceFromPictureElement()`:

```
ImageCandidate bestFitSourceForImageAttributes(float deviceScaleFactor, const AtomString& srcAttribute, StringView srcsetAttribute, float sourceSize, Function<bool(const ImageCandidate&)>&& shouldIgnoreCandidateCallback)
{
    if (srcsetAttribute.isNull()) {
```

I also put logging here, and verified that we reach it (with the callback getting called) if I modify my test case to use `<picture><source …><img …></picture>`

(Perhaps there's something subtle I'm missing, though?)

shallawa commented on Source/WebCore/html/parser/HTMLSrcsetParser.cpp:
```diff
@@ -314,6 +314,9 @@ ImageCandidate bestFitSourceForImageAttributes(float deviceScaleFactor, const At
     if (!srcAttribute.isEmpty())
         imageCandidates.append(ImageCandidate(StringViewWithUnderlyingString(srcAttribute, srcAttribute), DescriptorParsingResult(), ImageCandidate::SrcOrigin));
 
+    if (shouldIgnoreCandidate)
```
I think `shouldIgnoreCandidate` is a little bit confusing. Can it indicate it is a callback? Like `ignoreCandidateFunction`,  `ignoreCandidateMatch` `shouldIgnoreCandidateCallback` or something similar.

whsieh replied:
`shouldIgnoreCandidateCallback` sounds good to me!

shallawa commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -677,4 +678,27 @@ VisibleInViewportState ImageLoader::imageVisibleInViewport(const Document& docum
+        return false;
+
+    RefPtr page = document->protectedPage();
+    return page && !page->allowsLoadFromURL(candidateURL, MainFrameMainResource::No);
```
If `page` is null, is it right to keep this candidate and load it from the network? Should not this condition be:
```
    return !page || !page-> allowsLoadFromURL(...);
```

whsieh replied:
Hm.. I think that in practice, it doesn't really matter — I only did it this way to preserve existing behavior in case page is null. Will change to `return !page || !page-> allowsLoadFromURL(...);`

shallawa commented on Tools/TestWebKitAPI/Tests/mac/LoadWebArchive.mm:
```diff
@@ -644,4 +646,46 @@ HTTPServer server({
+    auto naturalWidth = dimensions.firstObject.doubleValue;
+    auto naturalHeight = dimensions.lastObject.doubleValue;
+    EXPECT_GT(naturalWidth, 0);
+    EXPECT_GT(naturalHeight, 0);
```
I think these conditions will be always true even for broken images. I think they will be { 20, 20 } if the default broken image is drawn. Should not these be { 400, 400 }?

whsieh replied:
I verified that without this fix, the natural width and height are indeed 0 (so the test fails).

However, I can adjust the test to make this check clearer and use `EXPECT_EQ` instead.

szewai replied:
My memory is broken image is used for image load failure, and with the filtering the image load will not even start

whsieh replied:
FWIW, the load failure that results in the broken image comes from this codepath:

```
    if (auto* page = frame ? frame->page() : nullptr) {
        auto mainFrameMainResource = frame->isMainFrame()
            && resourceLoader.frameLoader()
            && resourceLoader.frameLoader()->notifier().isInitialRequestIdentifier(identifier)
            ? MainFrameMainResource::Yes : MainFrameMainResource::No;
        if (!page->allowsLoadFromURL(request.url(), mainFrameMainResource)) {
            RunLoop::main().dispatch([resourceLoader = Ref { resourceLoader }, error = blockedError(request)] {
                resourceLoader->didFail(error);
            });
            return;
        }
    }
```

It's true that the image load never actually starts (in the network process, at least) — but this does seem to result in a broken image being shown still.

whsieh committed:
QuickLook sometimes shows broken images for image sources when loading web archives
https://bugs.webkit.org/show_bug.cgi?id=272005
rdar://123918480

Reviewed by Sihui Liu.

When previewing web archives, QuickLook (more precisely, `QLPreviewController`) loads the archive
data in a `WKWebView` that is configured to have an empty set of `_allowedNetworkHosts` (thereby
preventing network access when loading the archive). While this is a perfectly reasonable behavior
for the purposes of loading web archives, it exposes an issue when rendering images with sources
specified via the `srcset` attribute — if the "best fit" chosen for the image in QuickLook's web
view isn't the same source as the one that was chosen when creating the web archive, we end up with
a broken image since the selected source isn't present in the web archive, and we also can't
download it over the network. Since everything from viewport/screen dimensions to the device scale
factor can influence this, we can end up with broken images when copying images with `srcset`s in
many cases.

To mitigate this, we add some logic to detect when image sources are unavailable when loading web
archives (*and* we're additionally not allowed to fetch them from over the internet), and avoid
surfacing these image sources as potential candidates when selecting the best fit. See below for
more details.

Test: LoadWebArchive.PreferCachedImageSourceOverBrokenImage

* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::bestFitSourceFromPictureElement):
(WebCore::HTMLImageElement::selectImageSource):

Pass the new filtering function (using `ImageLoader::shouldIgnoreCandidateWhenLoadingFromArchive`
below) into `bestFitSourceForImageAttributes`.

* Source/WebCore/html/parser/HTMLSrcsetParser.cpp:
(WebCore::bestFitSourceForImageAttributes):

Add a new optional filtering function that takes an `ImageCandidate` and returns whether or not we
can even attempt to load the candidate; for now, this is `false` only in the case where we're
loading archived resources, and we're not allowed to load the URL over the network.

* Source/WebCore/html/parser/HTMLSrcsetParser.h:
(WebCore::bestFitSourceForImageAttributes):
* Source/WebCore/loader/DocumentLoader.h:
(WebCore::DocumentLoader::hasArchiveResourceCollection const):

Add a new helper function, which we can use to bail early in
`ImageLoader::shouldIgnoreCandidateWhenLoadingFromArchive`, to avoid eagerly parsing all URLs in the
`srcset`.

* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::shouldIgnoreCandidateWhenLoadingFromArchive const):
* Source/WebCore/loader/ImageLoader.h:
* Tools/TestWebKitAPI/Tests/mac/LoadWebArchive.mm:
(TestWebKitAPI::TEST(LoadWebArchive, PreferCachedImageSourceOverBrokenImage)):

Add a new API test to exercise the change, by loading a web archive from a web view at 1x device
pixel scale that contains an image with a `srcset`, and loading the archive in another web view at
2x scale. Before the changes in this patch, this results in a broken image.

* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView contentsAsWebArchive]):

Canonical link: https://commits.webkit.org/276952@main
